runners/flink/src/main/java/org/apache/beam/runners/flink/execution/SingletonSdkHarnessManager.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.flink.execution;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import javax.annotation.Nullable;
+import org.apache.beam.model.fnexecution.v1.ProvisionApi;
+import org.apache.beam.model.pipeline.v1.RunnerApi;
+import org.apache.beam.runners.fnexecution.ServerFactory;
+import org.apache.beam.runners.fnexecution.artifact.ArtifactSource;
+
+/**
+ * A long-lived class to manage SDK harness container instances on behalf of
+ * shorter-lived Flink operators.  It is responsible for initializing and storing
+ * job resources for retrieval.
+ */
+public class SingletonSdkHarnessManager implements  SdkHarnessManager {
+  private static class LazyHolder {
+    static final SingletonSdkHarnessManager INSTANCE = new SingletonSdkHarnessManager();
+  }
+
+  /** Get the singleton instance of the harness manager. */
+  public static SingletonSdkHarnessManager getInstance() {
+    return LazyHolder.INSTANCE;
+  }
+
+  private final ServerFactory serverFactory;
+  private final ExecutorService executorService;
+  private final JobResourceManagerFactory jobResourceManagerFactory;
+
+  private SingletonSdkHarnessManager() {
+    this.serverFactory = ServerFactory.createDefault();
+
+    ThreadFactory threadFactory =
```

Comment:
Since the constructor is doing real work now, we should probably move this.

I think the best approach is to remove the empty-param constructor altogether and move this logic into a static factory method that calls the normal constructor. You can then initialize the lazy holder with the factory method (i.e., `INSTANCE = myFactory()`.

Reply:
So I went ahead and made this change, because it seemed rather simple.  At some point though I'd like to have a conversation about what makes this better.  I can't see any difference in execution outcomes here, and the code style seems sort of irrelevant given that these are fairly uncomplex class-private methods.

Reply:
Long story short: testability and ensuring that partially-initialized objects never exist. I realize the benefits are minimal here, but having the factory method for non-trivial construction should make it clear to future contributors that no work should be done in the constructor and makes it easier to keep it this way.

For more context, see http://misko.hevery.com/code-reviewers-guide/flaw-constructor-does-real-work/.